10/EPMDX protocol documentation

vtkhir opened the pull request.

Commit:
EPMDX-2: Added doc of metadata get operation

Commit:
EPMDX-2: Added full and partial metadata update operations

Commit:
EPMDX-2: Added delete key operation, added more details about `id` field

Commit:
EPMDX-2: Tune the example description

Commit:
Merge remote-tracking branch 'origin/master' into EPMDX-2

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+## Requirements and constraints
+
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
```
Let's start from metadata definition. At this point in the document, the reader has no clue what you mean by metadata.

vtkhir replied:
Added the definition of `metadata`

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
+All UTF-8 characters are supported in metadata key names.
```
This looks like a constraints list. Format it as a list (or separate paragraphs), or it will be squashed into a single paragraph.

vtkhir replied:
OK, I will update the format of this paragraph after discussion about empty metadata keys

vtkhir replied:
Updated the requirements to metadata keys

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+For example, `longitude`, `street`, `name`, `$address 2` are valid metadata keys.
+No duplicate metadata keys allowed.
+### Metadata values
+Valid metadata value type is any JSON type:
```
No need to list them.

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
+All UTF-8 characters are supported in metadata key names.
+Metadata key must contain at least one non-whitespace character.
```
Why?

vtkhir replied:
What is the point to allow blank strings as a metadata keys?

rasendubi replied:
I don't see any reason to forbid it. A random use case from top of my head is storing some info for a string prefix, so string "hi" can use the following prefixes:
- "hi"
- "h"
- "" (as a common prefix for all strings)

akokhanovskyi replied:
Making sure everyone is on the same page:
- empty *values* are permitted;
- empty *keys* are *not permitted*. I don't see a good use case for supporting an empty key. It makes no sense, especially due to the keys being unique.

rasendubi replied:
We have already discussed that with @vtkhir and we agreed to support empty/whitespace-only keys. The main reason is "why not?"

I mean, there is no reason to reject them as this is a benefit for neither implementation nor usage. (Uniqueness is not an argument here, because `""` could be unique.)

Furthermore, if you don't see a use case now, that is not the reason to forbid the usage as a user may come with a use case later and then they are doomed to implement hacks and workarounds.

Finally, I think this complicates implementations as there is an additional check now (which includes decoding Unicode strings?). Also, "whitespace" is not defined. Does it include spaces? tabs? newlines? carriage return? no-break space? All Unicode whitespaces?

I'd rather treat keys as opaque strings and wouldn't make any assumptions (except for uniqueness, of course).

akokhanovskyi replied:
I don't think anyone had said anything about not allowing whitespace :)
*Empty* strings are questionable, and I understand this **might** be a problem **if** someone someday comes up with a killer scenario that would totally depend on an empty key support.
@vtkhir @rasendubi I leave it to you, guys, to decide. I don't see why would one want an empty key in metadata. Makes no sense to me.

rasendubi replied:
> Metadata key must contain at least one non-whitespace character.

That means whitespace-only strings are not supported.

Also, schema below defines keys as `\\S+`, which is "1+ non-space characters" so _no whitespaces allowed at all_.

rasendubi replied:
Anyway, @vtkhir could you please confirm we reach consensus here?

vtkhir replied:
@rasendubi, I think we should not allow the metadata keys without non-whitespace characters at least for the two reasons:

- Display such key-value pairs on UI in endpoint metadata widget will be a real problem. 
- Display such key-value pairs in EPMDX service instance configuration(list of readable and writable keys) on UI will be the same problem. For example, how to distinguish an empty string and string with one space?
- REST API for the EPR service now has a `PUT api/v1/endpoints/{endpointId}/metadata/{metadataKey}` method to update a single endpoint metadata key
If we will allow the empty metadata key, then the request will look like `PUT api/v1/endpoints/{endpointId}/metadata//`. It is really confusing for me.
- EPMDX and EPR primary designed to store a real metadata like `location`, `name`, `device ID`, etc. It makes no sense to build a [trie](https://en.wikipedia.org/wiki/Trie) in the root of metadata.
We will be able to build a trie in the some metadata key:
```
{
"trie": 
  { "": "Root prefix",
    "p": "P prefix"
  },
"otherValue": 1
}
```

rasendubi replied:
A little correction: `PUT api/v1/endpoints/{endpointId}/metadata/{metadataKey}` with empty `metadataKey` is `api/v1/endpoints/{endpointId}/metadata/`, not `api/v1/endpoints/{endpointId}/metadata//`. Anyway, both look fine for me.

As for UI, we don't have a requirement to make all arbitrary keys distinguishable. With Unicode, you сan't do that anyway. (Have you noticed that `с` in `сan't` is actually Cyrillic?)

A simple possibility does not mean the user will create _dozens_ of whitespace-only keys, so I don't care.

Ok, so here is an opposite proposal. Rather than make arbitrary limitations, let's take ones from [NATS](https://nats.io/documentation/internals/nats-protocol/#nats-protocol-conventions) (but probably forbid dots):
> Subject names: Subject names, including reply subject (INBOX) names, are case-sensitive and must be non-empty alphanumeric strings with no embedded whitespace, and optionally token-delimited using the dot character (.)

These limitations are perfect for data storages of any kind (and basically any system). As a bonus, this would allow us to publish metadata updates within NATS on a per-key basis.

vtkhir replied:
> with empty metadataKey is api/v1/endpoints/{endpointId}/metadata/

In the EPR REST API we have two methods:

- `PUT api/v1/endpoints/{endpointId}/metadata` - updates all metadata
- `PUT api/v1/endpoints/{endpointId}/metadata/{metadataKey}` - updates single metadata key

So update of single metadata key URI should look exactly ` api/v1/endpoints/{endpointId}/metadata//`

> let's take ones from NATS (but probably forbid dots)

I agree on your proposal.
@akokhanovskyi what do you think about restriction of metadata key with following regex `^[a-zA-Z0-9]+$`?

rasendubi replied:
gnatsd allows any characters except space, horizontal tab, dot, and `>`, so we might want to relax this rule in the future.

vtkhir replied:
But [NATS protocol documentation](http://nats.io/documentation/internals/nats-protocol/) restricts subjects as you mentioned above:

> Subject names, including reply subject (INBOX) names, are case-sensitive and must be non-empty alphanumeric strings with no embedded whitespace, and optionally token-delimited using the dot character

How gnatsd can allow all characters ?

rasendubi replied:
Discussed locally. By gnatsd I mean a concrete implementation—I have checked gnatsd sources.

vtkhir replied:
I updated the requirements to metadata keys

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+### Supported operations
+
+EPMDX protocol must support the following operations:
+- Get metadata key-value pairs (all or for a certain keys)
```
`for certain keys`

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+
+### Full metadata update
+
+Create or update endpoint metadata operation.
```
What does this sentence mean?

vtkhir replied:
Updated the description of this operation

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+
+Create or update endpoint metadata operation.
+In case of update the provided metadata will override all existing metadata key-value pairs.
+Use [partial update operation](#partial-metadata-update) to avoid removing an existing values.
```
Still not sure what this section is about.

rasendubi replied:
Ok. I got it know.

I think this section should have more wording.

vtkhir replied:
Updated the description in this section

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+
+  `<endpoint_token>/update`
+
+  where `endpoint_token` identifies the endpoint.
```
```
where `<endpoint_token>` identifies the endpoint.
```

vtkhir replied:
In the 5/CMX we have `endpoint_token` without angle brackets. I think we should keep the same style. Not sure which is better

rasendubi replied:
If you replace `endpoint_token` exactly, you'll get something like `<ic5Eo7ohsh>/update`, which is not what expected. The whole expression including angle brackets should be replaced.

rasendubi replied:
Also, both 1/KP and 3/MIPC mention brackets (both `<>` and `{}`) as replaceable parts.

vtkhir replied:
Ok, I'll update this RFC and `5/CMX`

vtkhir replied:
Updated current RFC and 5/CMX

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+      "type": "object",
+      "minProperties": 1,
+      "patternProperties": {
+        "\\S+": {
```
According to JSON schema, you can use `{}` as a catch-all type. (Instead of listing all types.)

vtkhir replied:
done

rasendubi reviewed (commented):
haven't finished a full review yet. (publishing my intermediate comments)

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+
+Client MUST send requests with the following extension-specific [resource path](/0001-kaa-protocol/README.md#resource-path-format) part in order to receive the metadata key-value pairs:
+
+  `<endpoint_token>/update/keys`
```
I propose renaming full and partial updates to `/replace` and `/update` respectively.

vtkhir replied:
The `replace` path may be confusing because the replace operation is supposed to be executed on already existing metadata key-value pair. But in partial update we support `create or update` of key-value pair. 
I'm thinking on other thing: do we have a use-case when endpoint should perform full update instead of partial?

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,687 @@
+The request payload MUST be a JSON-encoded object with the same [JSON schema](http://json-schema.org/) as in [full metadata update request](#full-metadata-update-request).
+In the `metadata` JSON property specify only key-value pairs that need to be updated.
+
+If there is no `id` JSON property the request payload than server will process the request but will not publish the status response back to client.
```
s/than/then/

rasendubi replied:
_in_ the request payload.

rasendubi replied:
s/will/MUST/

Commit:
EPMDX-2: Fixed typos and JSON schemas

Commit:
EPMDX-2: Doc fixes according to review comments

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,658 @@
+The metadata provides information about the endpoint and is a collection of key-value pairs.
+Example endpoint's metadata keys are `name`, `description`, `location`, `vendor`, `device model`, `firmware version` etc.
+
+## Requirements and constraints
```
This section mostly lists requirements _created by_ this protocol, not requirements _to_ the protocol.

vtkhir replied:
OK, so should I remove the list of supported operations from this section and change this title to `Constraints`?

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,658 @@
+- Get metadata key-value pairs (all or for certain keys)
+- Get a list of metadata keys
+- Update metadata key-value pairs (all or partial update)
+- Delete a single metadata key value pair by key name
```
Why single?

vtkhir replied:
Because currently we only support a single key delete operation

vtkhir replied:
Updated the delete operation to support multiple keys

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
````diff
@@ -0,0 +1,658 @@
+}
+```
+Examples:
+- get metadata without message identifier
````
That means the client will not receive metadata. This does not make sense for `get metadata` operation.

vtkhir replied:
Currently the message identifier in the get request from client is optional and client receives response from server in any case. Should we make it required in the client request?

**Note**: for update/delete operations the message id is also optional and client will not receive the response from server if `"id"` was not present in the request

vtkhir replied:
will create a separate story on EPMDX to match the current RFC

rasendubi commented on 0010-endpoint-metadata-extension/README.md:
```diff
@@ -0,0 +1,658 @@
+      "type": "integer",
+      "description": "The message identifier used to match server response to the request"
+    },
+    "keyName": {
```
Why not `key`? Or even better, `keys`

vtkhir replied:
Currently we support only deletion of one key, so `keys` is not suitable name. We can rename this field to `key`

vtkhir replied:
Renamed to `keys`

rasendubi committed:
10/EPMDX Review

The resource path is a slightly different beast than MQTT topic
name. Resource path is defined in 1/KP and is an abstraction for
hiding MQTT/CoAP differences. When we talk about request/response, the
resource path is the same for both requests and responses, MQTT topic
names are different, and CoAP `URI-Path`s are equal. `/status`
appending is MQTT-specific and 1/KP extensions should be independ from
underlying 1/KP protocol, so I removed that part.

Recent additions standardize the meaning of request `id` field, so I
adapt this protocol to the new conventions.

Also, fix some grammar, remove TOC (as it will be auto-generated), and
add 10/EPMDX to the RFC index.

rasendubi reviewed (commented):
I also made a PR with recommended changes at https://github.com/vtkhir/kaa-rfcs/pull/1

Commit:
Merge pull request #1 from rasendubi/10-epmdx

10/EPMDX Review

rasendubi committed:
Merge branch 'master' into EPMDX-2

Commit:
EPMDX-2: Fixed the partial update operation paragraph

Commit:
EPMDX-2: Restricted the metadata keys to be alpha-numeric strings

Commit:
EPMDX-2: Add support of delete multiple keys operation

Commit:
EPMDX-2: Updated constraints for endpoint metadata keys

Commit:
EPMDX-2: Fixed responses description

rasendubi committed:
10/EPMDX Review

"Metadata keys" and "Metadata values" sections are not initial
requirements to the protocol and are a part of the design. Move them
to the "Design" section.

Add the reasoning behind the metadata key format.

Specify JSON encoding (UTF-8) and rephrase the sentence to identify
which file it refers to.

Add myself to contributors. Change Volodymyr's address to the new
domain.

rasendubi reviewed (commented):
Opened a PR with my recommended changes https://github.com/vtkhir/kaa-rfcs/pull/2

Commit:
Merge pull request #2 from rasendubi/cr/56

10/EPMDX Review
